### PR TITLE
cleaned up suffix generation for CloudNames and default name for CasC

### PR DESCRIPTION
**Pull request by vineeth-bandi**
Suffix generation now uses a randomly generated string instead of incremental for default cloud name generation through UI or empty strings in CasC configurations.

### Testing done
Altered and added Unit Tests in [EC2CloudNamesTest.java](https://github.com/vineeth-bandi/ec2-fleet-plugin/blob/suffix-generation/src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java) for new functionality.

Added Unit Test in [EC2FleetCloudConfigurationAsCodeTest.java](https://github.com/vineeth-bandi/ec2-fleet-plugin/blob/suffix-generation/src/test/java/com/amazon/jenkins/ec2fleet/EC2FleetCloudConfigurationAsCodeTest.java) for empty strings in CasC.

Tested that both UI and CasC generate default cloud names properly. On creation of new clouds and empty strings in yaml file for UI and CasC, respectively.

Verified functionality of added features in Jenkins version 2.277.2 and version 2.414

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Collectors;
 
 public class CloudNames {
 
+  public static final int SUFFIX_LENGTH = 8;
+
+  public static Set<String> usedSuffixes = new HashSet<String>();
```
Comment (vineeth-bandi) [inline src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:15]: We could potentially do without this change since the probability of a repeat cloud name is $\frac{x}{62^{y}}$ where x is the total number of clouds constructed and y is `SUFFIX_LENGTH`. Even if x is 1 billion thats only a 0.0004% chance of getting a collision.

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java [hunk 2]
```diff
@@ -18,11 +24,29 @@ public static String generateUnique(final String defaultName) {
...
-    int suffix = 1;
     while (usedNames.contains(uniqName)) {
-      uniqName = defaultName + "-" + suffix++;
+      uniqName = defaultName + "-" + generateSuffix();
     }
 
     return uniqName;
   }
+
+  public static String processCloudName(String name) {
```
Comment (vineeth-bandi) [inline src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:34]: We can combine this and generateUniqueCloud name but I left them separate in case we eventually need some other form of name processing. If we don't think this is necessary I will go ahead and combine them.

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java [hunk 3]
```diff
@@ -18,11 +24,29 @@ public static String generateUnique(final String defaultName) {
...
+    if (name.equals("")){
+      return EC2FleetCloud.DEFAULT_FLEET_CLOUD_ID + "-" + generateSuffix();
+    }
+
+    return name;
+  }
+
+  private static String generateSuffix() {
```
Comment (pdk27) [inline src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:42]: Could you add some context here about why we need this?

File: src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java [hunk 4]
```diff
@@ -313,6 +313,10 @@ public int getMinSpareSize() {
         return minSpareSize;
     }
 
+    public int getMaxTotalUses() {
```
Comment (pdk27) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java:316]: For clarity, can you separate this fix into its own PR?

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java (hunk 2 above)
Comment (pdk27) [reply src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:34]: Yea! I think it makes sense to just use `generateUnique` in the Cloud constructors. Left a suggestion.

File: src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java [hunk 5]
```diff
@@ -188,7 +188,7 @@ public EC2FleetCloud(@Nonnull final String name,
                          final boolean scaleExecutorsByWeight,
                          final Integer cloudStatusIntervalSec,
                          final boolean noDelayProvision) {
-        super(name);
+        super(CloudNames.processCloudName(name));
```
Comment (pdk27) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java:191]:
```suggestion
        super(StringUtils.isNotBlank(name) ? name : CloudNames.generateUnique(DEFAULT_FLEET_CLOUD_ID));
```

File: src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetLabelCloud.java [hunk 6]
```diff
@@ -125,7 +125,7 @@ public EC2FleetLabelCloud(final String name,
                               final Integer cloudStatusIntervalSec,
                               final boolean noDelayProvision,
                               final String ec2KeyPairName) {
-        super(name);
+        super(CloudNames.processCloudName(name));
```
Comment (pdk27) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetLabelCloud.java:128]:
```suggestion
        super(StringUtils.isNotBlank(name) ? name : CloudNames.generateUnique(DEFAULT_FLEET_CLOUD_ID));
```

File: src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java [hunk 7]
```diff
@@ -60,6 +60,31 @@ public void generateUnique_addsSuffixCorrectly() {
             "-1", false, 0, 0, false,
             10, false));
 
-    Assert.assertEquals("FleetCloud-2", CloudNames.generateUnique("FleetCloud"));
+    String actual = CloudNames.generateUnique("FleetCloud");
+    Assert.assertTrue(actual.length() == ("FleetCloud".length() + CloudNames.SUFFIX_LENGTH + 1));
+    Assert.assertTrue(CloudNames.usedSuffixes.contains(actual.substring("FleetCloud".length()+1)));
```
Comment (pdk27) [inline src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java:65]: I think using `StringUtils.substringAfter` will be cleaner.

File: src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java [hunk 8]
```diff
@@ -60,6 +60,31 @@ public void generateUnique_addsSuffixCorrectly() {
...
+  @Test
+  public void processCloudName_emptyString() {
+    EC2FleetCloud fleetCloud = new EC2FleetCloud("", null, null, null, null, null,
+            "test-label", null, null, false, false,
+            0, 0, 0, 0, 0, true, false,
+            "-1", false, 0, 0, false,
+            10, false);
+
+    Assert.assertTrue(fleetCloud.name.length() == ("FleetCloud".length() + CloudNames.SUFFIX_LENGTH + 1));
+    Assert.assertTrue(fleetCloud.name.contains("FleetCloud-"));
```
Comment (pdk27) [inline src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java:77]: `startsWith`  instead of `contains`?

**Review by pdk27 — CHANGES_REQUESTED** [review]

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java [hunk 9]
```diff
@@ -1,28 +1,49 @@
...
+    if (proposedName.equals(EC2FleetCloud.DEFAULT_FLEET_CLOUD_ID) || proposedName.equals(EC2FleetLabelCloud.DEFAULT_FLEET_CLOUD_ID) || usedNames.contains(proposedName)) {
+      return proposedName + "-" + generateSuffix();
     }
 
-    return uniqName;
+    return proposedName;
+  }
+  
+  /**
+   * We are using a randomly generated string as a suffix here because Jenkins recreates clouds on config change.
```
Comment (pdk27) [inline src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:34]: Adding some more context: `Jenkins recreates clouds on config change and makes the new clouds accessible only after all clouds are replaced which makes identifying clouds with duplicate names during (re-)creation impossible, esp. for JCasC` ?

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java [hunk 10]
```diff
@@ -1,28 +1,49 @@
...
+  public static String generateUnique(final String proposedName) {
     final Set<String> usedNames = Jenkins.get().clouds != null
             ? Jenkins.get().clouds.stream().map(c -> c.name).collect(Collectors.toSet())
             : Collections.emptySet();
-    String uniqName = defaultName;
 
-    int suffix = 1;
-    while (usedNames.contains(uniqName)) {
-      uniqName = defaultName + "-" + suffix++;
+    if (proposedName.equals(EC2FleetCloud.DEFAULT_FLEET_CLOUD_ID) || proposedName.equals(EC2FleetLabelCloud.DEFAULT_FLEET_CLOUD_ID) || usedNames.contains(proposedName)) {
```
Comment (pdk27) [inline src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:26]: We can safely omit the proposeName checks here by allowing 1 cloud in default name for each cloud type, can't we?

File: src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java [hunk 11]
```diff
@@ -33,33 +34,58 @@ public void isUnique_false() {
 
   @Test
   public void generateUnique_noSuffix() {
-    Assert.assertEquals("FleetCloud", CloudNames.generateUnique("FleetCloud"));
+    Assert.assertEquals("UniqueCloud", CloudNames.generateUnique("UniqueCloud"));
```
Comment (pdk27) [inline src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java:37]: I guess if `proposedName` checks are removed in, then there is no need to rename the clouds in these tests?

File: src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java [hunk 12]
```diff
@@ -33,33 +34,58 @@ public void isUnique_false() {
...
+
+  @Test
+  public void generateUnique_emptyStringInConstructor() {
+    EC2FleetCloud fleetCloud = new EC2FleetCloud("", null, null, null, null, null,
+            "test-label", null, null, false, false,
+            0, 0, 0, 0, 0, true, false,
+            "-1", false, 0, 0, false,
+            10, false);
+
+    Assert.assertTrue(fleetCloud.name.length() == ("FleetCloud".length() + CloudNames.SUFFIX_LENGTH + 1));
```
Comment (pdk27) [inline src/test/java/com/amazon/jenkins/ec2fleet/CloudNamesTest.java:77]: use `assertEquals` for better readability.

**Review by pdk27 — CHANGES_REQUESTED** [review]

**Comment by pdk27** [comment]
Could you add a test to https://github.com/jenkinsci/ec2-fleet-plugin/blob/master/src/test/java/com/amazon/jenkins/ec2fleet/EC2FleetCloudConfigurationAsCodeTest.java for the default name case?

File: src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java (hunk 10 above)
Comment (vineeth-bandi) [reply src/main/java/com/amazon/jenkins/ec2fleet/CloudNames.java:26]: Unfortunately not if a CasC user has multiple clouds with default name.

**Review by pdk27 — APPROVED** [review]
LGTM 🚀
